src/auth/Crypto.cc

```diff
@@ -348,24 +360,695 @@ int CryptoAES::validate_secret(const bufferptr& secret)
+      const bufferptr& iv,
+      string& error) override;
+
+  static constexpr size_t KEY_SIZE = 256 / 8;
+  static constexpr size_t IV_SIZE = 128 / 8;
+  static constexpr size_t BLOCK_SIZE = 128 / 8;
+};
+
+
+class CryptoAES_256_CTR : public CryptoHandler {
+public:
+  CryptoAES_256_CTR() { }
+  ~CryptoAES_256_CTR() override {}
+  int get_type() const override {
+    return CEPH_CRYPTO_AES_256_CTR;
+  }
+  int create(bufferptr& secret) override {
+    char d[KEY_SIZE];
+    bufferptr p(KEY_SIZE);
+    int r = get_random_bytes(d, KEY_SIZE);
+    if (r < 0)
+      return r;
+    secret = buffer::ptr(d, KEY_SIZE);
+    return 0;
+  }
+  int validate_secret(const bufferptr& secret) override {
+    if (secret.length() != KEY_SIZE) {
+      return -EINVAL;
+    }
+    return 0;
+  }
+  CryptoKeyHandler *get_key_handler(
+      const bufferptr& secret,
+      string& error) override;
+  CryptoKeyHandler *get_key_handler(
+      const bufferptr& secret,
+      const bufferptr& iv,
+      string& error) override;
+
+  static constexpr size_t KEY_SIZE = 256 / 8;
+  static constexpr size_t IV_SIZE = 128 / 8;
+  static constexpr size_t BLOCK_SIZE = 128 / 8;
+};
+
+
+class CryptoAES_256_ECB : public CryptoHandler {
+public:
+  CryptoAES_256_ECB() { }
+  ~CryptoAES_256_ECB() override {}
+  int get_type() const override {
+    return CEPH_CRYPTO_AES_256_ECB;
+  }
+  int create(bufferptr& secret) override {
+    char d[KEY_SIZE];
+    bufferptr p(KEY_SIZE);
+    int r = get_random_bytes(d, KEY_SIZE);
+    if (r < 0)
+      return r;
+    secret = buffer::ptr(d, KEY_SIZE);
+    return 0;
+  }
+  int validate_secret(const bufferptr& secret) override {
+    if (secret.length() != KEY_SIZE) {
+      return -EINVAL;
+    }
+    return 0;
+  }
+  CryptoKeyHandler *get_key_handler(
+      const bufferptr& secret,
+      string& error) override;
+  CryptoKeyHandler *get_key_handler(
+      const bufferptr& secret,
+      const bufferptr& iv,
+      string& error) override;
+
+  static constexpr size_t KEY_SIZE = 256 / 8;
+  static constexpr size_t BLOCK_SIZE = 128 / 8;
+};
+
+
+class CryptoAES_256_CBCKeyHandler : public CryptoKeyHandler {
+public:
+  CryptoAES_256_CBCKeyHandler(
+      const bufferptr& key,
+      const bufferptr& iv)
+    : iv(iv) {
+    secret = key;
+  }
+  virtual ~CryptoAES_256_CBCKeyHandler() {}
+
+  int encrypt(
+      const bufferlist& in,
+      bufferlist& out,
+      std::string *error) const {
+    return cbc_transform(in, out, true, error);
+  }
+  int decrypt(
+      const bufferlist& in,
+      bufferlist& out,
+      std::string *error) const {
+    return cbc_transform(in, out, false, error);
+  }
+private:
+  bufferptr iv;
+
+  int cbc_transform(
+      const bufferlist& in,
+      bufferlist& out,
+      bool encrypt,
+      std::string *error) const {
+    if ((in.length() % CryptoAES_256_CBC::BLOCK_SIZE) != 0) {
+      if (error) {
+        *error = "Input length not multiple of 16.";
+      }
+      return -EINVAL;
```

Comment:
the existing `CryptoAESKeyHandler` doesn't enforce block boundaries like this. can you help me understand why this is required for the new handlers? can we find a way to make the interface more explicit about this requirement (documentation or otherwise)?